opm/autodiff/AquiferCarterTracy.hpp

```diff
@@ -0,0 +1,418 @@
+                {
+                    auto cellFacesRange = cell2Faces[cell_idx_.at(idx)];
+                    
+                    for(auto cellFaceIter = cellFacesRange.begin(); cellFaceIter != cellFacesRange.end(); ++cellFaceIter)
+                    {
+                        // The index of the face in the compressed grid
+                        const int faceIdx = *cellFaceIter;
+
+                        // the logically-Cartesian direction of the face
+                        const int faceTag = Opm::UgGridHelpers::faceTag(ugrid, cellFaceIter);
+
+                        
+                        if (faceTag == 0) // left
+                            faceDirection = Opm::FaceDir::XMinus;
+                        else if (faceTag == 1) // right
+                            faceDirection = Opm::FaceDir::XPlus;
+                        else if (faceTag == 2) // back
+                            faceDirection = Opm::FaceDir::YMinus;
+                        else if (faceTag == 3) // front
+                            faceDirection = Opm::FaceDir::YPlus;
+                        else if (faceTag == 4) // bottom
+                            faceDirection = Opm::FaceDir::ZMinus;
+                        else if (faceTag == 5) // top
+                            faceDirection = Opm::FaceDir::ZPlus;
+
+                        if (faceDirection == connection.reservoir_face_dir.at(idx))
+                        {
+                            // Check now if the face is outside of the reservoir, or if it adjoins an inactive cell
+                            // Do not make the connection if the product of the two cellIdx > 0. This is because the
+                            // face is within the reservoir/not connected to boundary. (We still have yet to check for inactive cell adjoining)
+                            faceArea = (faceCells(faceIdx,0)*faceCells(faceIdx,1) > 0)? 0. : Opm::UgGridHelpers::faceArea(ugrid, faceIdx);
+                            faceArea_connected_.at(idx) = faceArea;
+                            denom_face_areas += ( connection.influx_multiplier.at(idx) * faceArea_connected_.at(idx) );
+                        }
+                    }
+                    auto cellCenter = grid.getCellCenter(cell_idx_.at(idx));
+                    cell_depth_.at(idx) = cellCenter[2];
+                }
+
+                for (size_t idx = 0; idx < cell_idx_.size(); ++idx)
+                {
+                    alphai_.at(idx) = ( connection.influx_multiplier.at(idx) * faceArea_connected_.at(idx) )/denom_face_areas;
+                }
+            }
+
+            inline void calculate_aquifer_condition()
+            {
+
+                int pvttableIdx = pvttableID_ - 1;
+                
+                rhow_.resize(cell_idx_.size(),0.);
+                
+                if (p0_defaulted_)
+                {
+                   pa0_ = calculate_reservoir_equilibrium(rhow_);
```

Comment:
rhoW_ is a member variable, it does not need to be an argument of the member function.